docs/versioned_docs/version-zenith/developer-guide/fundamentals/942201-execution-environment.mdx

```diff
@@ -0,0 +1,43 @@
+title: "Execution Environment"
+---
+
+import PartialExperimentalDocs from '../../partials/_experimental.mdx';
+
+# Execution Environment
+
+<PartialExperimentalDocs />
+
+## Containerization
+
+All Modules and their Functions execute within containers spawned by the Dagger Engine. This serves a few purposes:
+1. Reproducibility - executing in a well-defined and well-controlled container makes it easier for your code to run the same way every time it is invoked. This makes it easier to reason about; you can't accidentally create "hidden dependencies" on ambient properties of the execution environment that could change at any moment.
+1. Caching - the reproducible containerized environment makes it possible to cache the result of Function execution, which in turn allows Dagger to automatically speed up Function execution.
+1. Security - Even when running third-party Modules sourced from a Git repository, those Functions will not have default access to your host environment (e.g. files, directories, environment variables, etc.). They can only be granted access by explicitly passing those resources as argument values to the Function.
+
+## Code execution
+When implementing Functions, you are free to write arbitrary code that will execute inside the Module's container. You have access to the Dagger API to make calls to the Core Dagger API or other Modules you depend on, but you are also free to just use the language's standard library and/or imported third-party libraries.
+
+The process your code executes in will currently be with the "root" user, but without a full set of Linux capabilities and other standard container sandboxing provided by runC.
+
+The current working directory of your code will be an initially empty directory. You can write and read files and directories in this directory if needed. This includes using the `Export` APIs on `File`/`Directory`/`Container` to write those to this local directory if needed.
+
+## `CurrentModule` API
+The Dagger API client made available to all Functions includes a `CurrentModule` API that provides abilities to introspect the Function's Module and interface between the current execution environment and the Dagger API.
+
+For example:
+1. The `Source` API on `CurrentModule` returns a `Directory` where your Module's source code is stored, which grants access to any files or directories within there.
+1. The `Workdir` and `WorkdirFile` APIs on `CurrentModule` allow you to load directories and files, respectively, from your Functions current working directory (as described in the above section).
+
+# Using Host Resources
+
+As mentioned above, Functions execute in containers and thus do not have access to resources on the host you invoke the Function from (i.e. the host you execute a CLI command like `dagger call` from). Instead, these resources need to be explicitly passed when executing `dagger call`.
+
+This includes: (TODO:(sipsma) include code examples here? Or just link somewhere else?)
```

Comment:
cc @vikram-dagger were you imagining these docs in `fundamentals` to include CLI examples too? Or should these be super high-level and just link to other sections? I'm happy to fill in other sections with any details mentioned here that are missing if so. Or can also just provide full examples here too. Just let me know what you think.

Reply:
Yes, of course, you can add CLI examples if you like. However for the specific items you have here, we already have various examples (see Arguments page for an example of passing a host/remote dir, Secrets page for passing an env var, Chaining and Quickstart pages for chaining examples, etc). So it's fine to link to those pages if easier. We don't yet have a Services page with an example, mainly due to time constraints so that link can be added later once a page is ready.

Reply:
Thanks, updated to link to the "Pass argument to a function" guide here too!